### PR TITLE
Bugfix encryption distinct select bug

### LINKED ISSUE #20096: encryption distinct select bug~
## Bug Report
Mysql 5.7 with shardingsphere 5.1.2
-- `sharding-demo`.t_user definition

CREATE TABLE `t_user` (
  `user_id` bigint(20) NOT NULL COMMENT '用户唯一ID',
  `fullname` varchar(50) DEFAULT NULL COMMENT '名字',
  `user_type` varchar(255) DEFAULT NULL COMMENT '类型',
  `cipher_pwd` varchar(255) DEFAULT NULL COMMENT '密码',
  `mobile` varchar(100) DEFAULT NULL COMMENT '手机号',
  `mobile_data` varchar(100) DEFAULT NULL COMMENT '手机号',
  `id_card` varchar(60) DEFAULT NULL COMMENT '身份证',
  PRIMARY KEY (`user_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;

I want the encryption fullname field,insert is OK, select is OK, bug use select distinct fullname will throw exception 

Exception in thread "main" java.lang.StringIndexOutOfBoundsException: String index out of range: -8
	at java.lang.String.substring(String.java:1967)
	at org.apache.shardingsphere.infra.rewrite.sql.impl.AbstractSQLBuilder.getConjunctionText(AbstractSQLBuilder.java:64)
	at org.apache.shardingsphere.infra.rewrite.sql.impl.AbstractSQLBuilder.toSQL(AbstractSQLBuilder.java:47)
	at org.apache.shardingsphere.infra.rewrite.engine.RouteSQLRewriteEngine.addSQLRewriteUnits(RouteSQLRewriteEngine.java:96)
	at org.apache.shardingsphere.infra.rewrite.engine.RouteSQLRewriteEngine.rewrite(RouteSQLRewriteEngine.java:72)
	at org.apache.shardingsphere.infra.rewrite.SQLRewriteEntry.rewrite(SQLRewriteEntry.java:73)
	at org.apache.shardingsphere.infra.context.kernel.KernelProcessor.rewrite(KernelProcessor.java:59)
	at org.apache.shardingsphere.infra.context.kernel.KernelProcessor.generateExecutionContext(KernelProcessor.java:47)
	at org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSphereStatement.createExecutionContext(ShardingSphereStatement.java:474)
	at org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSphereStatement.executeQuery(ShardingSphereStatement.java:156)

``` java
    val columnConfigAes =
        new EncryptColumnRuleConfiguration("fullname", "fullname", "", "", "name_encryptor", null);
 ....
  val query = connection.createStatement();
    ResultSet rs = query.executeQuery("select distinct t.fullname from t_user t ");
```

i debug with IDE found `AbstractSQLBuilder` toSQL() method getConjunctionText throw exception.
image

will be create 2 SQLToken with the save start stop index .

i use `select distinct t.fullname from t_user t order by fullname` is OK。
i have a question about `SelectStatementContext`,why create `OrderByContextEngine` Special handling `private OrderByContext createOrderByContextForDistinctRowWithoutGroupBy(final SelectStatement selectStatement, final GroupByContext groupByContext)`  distinct row must have order by expression？

### PR DESCRIPTION
**Pull request by wangtiga**
Fixes #20096.

Changes proposed in this pull request:
  - Bugfix encryption distinct select bug

---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [x] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: features/encrypt/core/src/main/java/org/apache/shardingsphere/encrypt/rewrite/token/generator/EncryptOrderByItemTokenGenerator.java [hunk 1]
```diff
@@ -101,7 +101,9 @@ private Collection<OrderByItem> getOrderByItems(final SQLStatementContext sqlSta
         }
         Collection<OrderByItem> result = new LinkedList<>();
         SelectStatementContext statementContext = (SelectStatementContext) sqlStatementContext;
-        result.addAll(statementContext.getOrderByContext().getItems());
+        if (!statementContext.getOrderByContext().isGenerated()) {
```
Comment (tuichenchuxin) [inline features/encrypt/core/src/main/java/org/apache/shardingsphere/encrypt/rewrite/token/generator/EncryptOrderByItemTokenGenerator.java:104]: Why we remove generated?

Comment (wangtiga) [reply features/encrypt/core/src/main/java/org/apache/shardingsphere/encrypt/rewrite/token/generator/EncryptOrderByItemTokenGenerator.java:104]: To address issue #4565, the sharding feature automatically adds the ORDER BY clause when processing the "SELECT distinct amount FROM t_account" statement. However, for the account_id column that has encryption rules configured, when processing the same statement, it only needs to rewrite the encrypted column name without adding the ORDER BY clause.

The presence of the orderByContext member variable in the SelectStatementContext class causes the problem described in issue #20096 when rewriting the SQL. It mistakenly assumes that the original SQL is "SELECT distinct amount FROM t_account ORDER BY amount".

References:
https://github.com/apache/shardingsphere/blob/5.3.2/test/it/rewriter/src/test/resources/scenario/sharding/case/dml/select.xml#L101
https://github.com/apache/shardingsphere/pull/4597

**Comment by strongduanmu** [comment]
Hi @wangtiga, can you merge master branch? I think this bug has been fixed.

```java
    private boolean containsOrderByItem(final SQLStatementContext sqlStatementContext) {
        if (!(sqlStatementContext instanceof SelectStatementContext)) {
            return false;
        }
        SelectStatementContext statementContext = (SelectStatementContext) sqlStatementContext;
        if (!statementContext.getOrderByContext().getItems().isEmpty() && !statementContext.getOrderByContext().isGenerated() || !statementContext.getGroupByContext().getItems().isEmpty()) {
            return true;
        }
        for (SelectStatementContext each : statementContext.getSubqueryContexts().values()) {
            if (containsOrderByItem(each)) {
                return true;
            }
        }
        return false;
    }
```

**Review by strongduanmu — APPROVED** [review]
Thank you @wangtiga.

**Comment by strongduanmu** [comment]
Hi @wangtiga, can you comment on the original issue? I will assign it to you.

**Pull request merged**